### PR TITLE
Fix file URIs on Windows

### PR DESCRIPTION
**Pull request by deivid-rodriguez**
### What was the end-user problem that led to this PR?

The problem was that file:// URIs were not properly handled on Windows, and those are heavily used by our specs.

### What was your diagnosis of the problem?

My diagnosis was that Windows paths do not start with a slash, so we need to add an extra slash to
separate the host from the path in file:// urls. Otherwise "D:" is parsed as the host segment in the URI.

As a consequence of that change, we also need to strip that leading character when using the path segment for those URIs, so that it's a valid file path on Windows.

This change reduces Windows CI spec failures from 429 to 355.

In order to apply this fix for all file URI's in tests, I extracted a helper method and searched and replaced all file:// ocurrences with the method. The diff is big because of that, but I think it works fine.

### CODE REVIEW

File: lib/bundler/fetcher/index.rb [hunk 1]
```diff
@@ -30,7 +30,8 @@ def fetch_spec(spec)
 
         uri = URI.parse("#{remote_uri}#{Gem::MARSHAL_SPEC_DIR}#{spec_file_name}.rz")
         if uri.scheme == "file"
-          Bundler.load_marshal Bundler.rubygems.inflate(Gem.read_binary(uri.path))
+          path = Gem.win_platform? ? uri.path[1..-1] : uri.path
```
Comment (ryanfox1985) [inline lib/bundler/fetcher/index.rb]: duplicated code, is it possible to avoid these duplication?

Comment (deivid-rodriguez) [reply lib/bundler/fetcher/index.rb]: Maybe, I'm not sure why this code is currently duplicated, but refactoring that is out of the scope of this PR.

Comment (janpio) [reply lib/bundler/fetcher/index.rb]: I think this comment referred specifically to the code `path = Gem.win_platform? ? uri.path[1..-1] : uri.path` you added - but as it seems the whole `fetch_spec` is duplicated this is indeed out of scope to extract this, considering that `fetch_spec` itself will probably be dedpulicated sooner or later (hopefully).

File: lib/bundler/source/rubygems/remote.rb [hunk 2]
```diff
@@ -25,8 +25,7 @@ def cache_slug
 
             cache_uri = original_uri || uri
 
-            # URI::File of Ruby 2.6 returns empty string when given "file://".
-            host = defined?(URI::File) && cache_uri.is_a?(URI::File) ? nil : cache_uri.host
+            host = cache_uri.to_s.start_with?("file://") ? nil : cache_uri.host
```
Comment (janpio) [inline lib/bundler/source/rubygems/remote.rb]: What is this change? Just a different way to achieve the same thing?

File: lib/bundler/fetcher/index.rb (hunk 1 above)
Comment (deivid-rodriguez) [reply lib/bundler/fetcher/index.rb]: Correct :+1:

File: spec/install/gemfile/sources_spec.rb [hunk 3]
```diff
@@ -588,26 +569,6 @@
           gem "rack-obama"
         G
 
-        # It creates this lockfile.
```
Comment (janpio) [inline spec/install/gemfile/sources_spec.rb]: unrelated change?

File: spec/install/gemfile/sources_spec.rb [hunk 4]
```diff
@@ -471,58 +471,39 @@
...
-        source 'file://#{gem_repo1}'
+        source '#{file_uri_for(gem_repo1)}'
         gem 'rack'
-        gem 'foo', '~> 0.1', :source => 'file://#{gem_repo4}'
-        gem 'bar', '~> 0.1', :source => 'file://#{gem_repo4}'
+        gem 'foo', '~> 0.1', :source => '#{file_uri_for(gem_repo4)}'
+        gem 'bar', '~> 0.1', :source => '#{file_uri_for(gem_repo4)}'
       G
 
-      # 3. this lockfile.
```
Comment (janpio) [inline spec/install/gemfile/sources_spec.rb]: unrelated change?

**Comment by janpio** [comment]
- Sometimes the exisiting code uses `file:`, sometimes `file://` - is this important?
- Is there something present to avoid "plain" file URLs being used again? Or would this be caught by a new test failure on Windows?

File: lib/bundler/source/rubygems/remote.rb (hunk 2 above)
Comment (deivid-rodriguez) [reply lib/bundler/source/rubygems/remote.rb]: Very good question.

My idea is to fix the issue that when `file://` URI is used with "localhost" (`file://localhost/path` instead of `file:///path`), then `localhost` would be prepended to the cache path, breaking [this spec](https://travis-ci.org/bundler/bundler/jobs/528808519).

Unfortunately I still need to use "localhost-type" file URIs for older rubies, because otherwise I can't seem to get a consistent behavior across versions (I guess because of bugs fixed in recent rubies):

https://github.com/bundler/bundler/blob/cdcde3e3e5d064dbcad7d20570de1732ba68ef9b/spec/support/path.rb#L69

The previous code use `file:/`style URI (single colon) to bypass this and get the spec passing:

https://github.com/bundler/bundler/blob/3d94636113db3f373be05514b7ae982db26b42e2/spec/install/global_cache_spec.rb#L197

But those don't work well on Windows.

The change makes sense to me since it makes the behavior consistent across ruby versions: we never include the host for `file://` URIs.

File: spec/install/gemfile/sources_spec.rb (hunk 3 above)
Comment (deivid-rodriguez) [reply spec/install/gemfile/sources_spec.rb]: Yeah, I can extract this changes to a separate PR. :+1:

**Comment by deivid-rodriguez** [comment]
> Sometimes the exisiting code uses file:, sometimes file:// - is this important?

The `file://` protocol specification is quite complicated, but in principle both are correct variants. I think the variability used in specs is because the standard library behaves quite differently accross rubies & platforms, I think most likely because of bugs fixed in recent versions. See https://github.com/bundler/bundler/pull/7161#discussion_r282456588.

> Is there something present to avoid "plain" file URLs being used again? Or would this be caught by a new test failure on Windows?

Can you clarify, I didn't understand what do you mean.

File: spec/install/gemfile/sources_spec.rb (hunk 3 above)
Comment (deivid-rodriguez) [reply spec/install/gemfile/sources_spec.rb]: Extracted to #7167. Will rebase this PR once that one is merged, so that these changes disappear from the diff.

**Comment by deivid-rodriguez** [comment]
As I understand the this, both `file:` and `file://` variants are correct. Then, if there's no host, we get a third slash meaning "absolute path starts". On Linux, this is the implicit behavior because absolute paths actually start with a slash, but on Windows it has to be added explicitly. Essentially, `file:///` URIs are correct, but the third slash is not part of the protocol.

**Comment by janpio** [comment]
> > Is there something present to avoid "plain" file URLs being used again? Or would this be caught by a new test failure on Windows?
> 
> Can you clarify, I didn't understand what do you mean.

I effectively answered this myself: If someone introduces a spec change that uses `file://` again without the new helper method, this spec will just fail on Windows CI. Correct?

**Comment by deivid-rodriguez** [comment]
> I effectively answered this myself: If someone introduces a spec change that uses file:// again without the new helper method, this spec will just fail on Windows CI. Correct?

Ah, I see. Yes, I'm pretty sure it would fail on Windows :+1:.

**Comment by deivid-rodriguez** [comment]
PR was rebased and now it doesn't have unrelated changes.

As a side note, I think this can probably be fixed in ruby-core. In my opinion, `ruby -e 'puts URI.parse("file:///c:/path/to/file").path'` should print `c:/path/to/file` instead of `/c:/path/to/file` as it does not.

**Comment by ESA-Security-Group** [comment]
Check

File: lib/bundler/fetcher.rb [hunk 5]
```diff
@@ -99,7 +99,8 @@ def fetch_spec(spec)
 
       uri = URI.parse("#{remote_uri}#{Gem::MARSHAL_SPEC_DIR}#{spec_file_name}.rz")
       if uri.scheme == "file"
-        Bundler.load_marshal Bundler.rubygems.inflate(Gem.read_binary(uri.path))
+        path = Gem.win_platform? ? uri.path[1..-1] : uri.path
```
Comment (deivid-rodriguez) [inline lib/bundler/fetcher.rb]: So, I wanted to make this logic a bit more robust (not always assume windows parsed paths start with `/`), and I found that this logic is already implemented in rubygems:

https://github.com/rubygems/rubygems/blob/d470f1787a94f751c907e8299ec44f61b7689432/lib/rubygems/remote_fetcher.rb#L320-L327

So I guess I can reuse that logic here?

Comment (deivid-rodriguez) [reply lib/bundler/fetcher.rb]: Opened rubygems/rubygems#2767 to make this easier.

Comment (deivid-rodriguez) [reply lib/bundler/fetcher.rb]: I included using the linked PR here :+1:

**Comment by deivid-rodriguez** [comment]
ping @hsbt since you assigned this to yourself :)

**Comment by deivid-rodriguez** [comment]
#7215 [failed](https://travis-ci.org/bundler/bundler/jobs/547819071) because of varying behaviour of `file://` URL parsing on different rubies. That should also be fixed by this PR.

**Comment by deivid-rodriguez** [comment]
Thanks @hsbt!

@bundlerbot r=hsbt

**Comment by ESA-Security-Group** [comment]
😂😂

Sincerely - 

Higher Commander - DPO
U.S AirForce - Armed Forces in Europe 
Unit: 527th - Space ࿇ Aggressor ࿇ Squadron

Ambjoern Skoeld

DPO16G4
Armed Forces of Europe - “EA”
APO/DPO Air/Army
 


19 juni 2019 kl. 18:38 skrev David Rodríguez <notifications@github.com>:

ping @hsbt since you assigned this to yourself :)
